RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoRU.cc

```diff
@@ -0,0 +1,464 @@
+std::vector<ME0Segment> ME0SegAlgoRU::run(const ME0Chamber * chamber, const HitAndPositionContainer& rechits){
+
+  #ifdef EDM_ML_DEBUG // have lines below only compiled when in debug mode
+  ME0DetId chId(chamber->id());
+  edm::LogVerbatim("ME0SegAlgoRU") << "[ME0SegmentAlgorithm::run] build segments in chamber " << chId << " which contains "<<rechits.size()<<" rechits";
+  for(unsigned int iH = 0; iH < rechits.size(); ++iH){
+	    auto me0id = rechits[iH].rh->me0Id();
+	    auto rhLP  = rechits[iH].lp;
+	    edm::LogVerbatim("ME0SegAlgoRU") << "[RecHit :: Loc x = "<<std::showpos<<std::setw(9)<<rhLP.x()<<" Glb y = "<<std::showpos<<std::setw(9)<<rhLP.y()<<" Time = "<<std::showpos<<rechits[iH].rh->tof()<<" -- "<<me0id.rawId()<<" = "<<me0id<<" ]"<<std::endl;
+  }
+  #endif
+
+
+  if (rechits.size() < 3 || rechits.size()>300){
+	  return std::vector<ME0Segment>();
+  }
+
+  theChamber = chamber;
+
+  std::vector<unsigned int> recHits_per_layer(6,0);
+  for(unsigned int iH = 0; iH < rechits.size(); ++iH) {
+	  recHits_per_layer[rechits[iH].layer-1]++;
+  }
+
+  BoolContainer used(rechits.size(),false);
+
+  // We have at least 2 hits. We intend to try all possible pairs of hits to start
+  // segment building. 'All possible' means each hit lies on different layers in the chamber.
+  // after all same size segs are build we get rid of the overcrossed segments using the chi2 criteria
+  // the hits from the segs that are left are marked as used and are not added to segs in future iterations
+  // the hits from 3p segs are marked as used separately in order to try to assamble them in longer segments
+  // in case there is a second pass
+
+  // Choose first hit (as close to IP as possible) h1 and second hit
+  // (as far from IP as possible) h2 To do this we iterate over hits
+  // in the chamber by layer - pick two layers.  Then we
+  // iterate over hits within each of these layers and pick h1 and h2
+  // these.  If they are 'close enough' we build an empty
+  // segment.  Then try adding hits to this segment.
+
+  std::vector<ME0Segment> segments;
+
+  auto doStd = [&] () {
+	  for(unsigned int n_seg_min = 6u; n_seg_min  >= stdParameters.minNumberOfHits; --n_seg_min)
+		  lookForSegments(stdParameters,n_seg_min,rechits,recHits_per_layer, used, segments);
+  };
+  auto doDisplaced = [&] () {
+	  for(unsigned int n_seg_min = 6u; n_seg_min  >= displacedParameters.minNumberOfHits; --n_seg_min)
+		  lookForSegments(displacedParameters,n_seg_min,rechits,recHits_per_layer, used,segments);
+  };
+  auto doWide = [&] () {
+	  for(unsigned int n_seg_min = 6u; n_seg_min >= wideParameters.minNumberOfHits; --n_seg_min)
+		  lookForSegments(wideParameters,n_seg_min,rechits,recHits_per_layer, used,segments);
+  };
+  auto printSegments = [&] {
+#ifdef EDM_ML_DEBUG // have lines below only compiled when in debug mode
+std::vector<ME0Segment>::iterator it =segments.begin();
+for(std::vector<ME0Segment>::iterator it =segments.begin(); it != segments.end(); ++it) {

```

Comment:
It depends on PU/neturon background. But you can get ~40 hits per chamber at nominal HL-LHC conditions. We get ~10 segments per chamber.

Anyways, Ill change it.